feat: add linting documentation

matthew-chirgwin opened the pull request:
- adds first version of linting documentation, along with proposed tools
to use, as well as some suggested rules to follow for UI code

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

matthew-chirgwin committed:
feat: add linting documentation

- adds first version of linting documentation, along with proposed tools
to use, as well as some suggested rules to follow for UI code

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

JQrdan commented on docs/Linting.md:
```diff
@@ -0,0 +1,82 @@
+
+to run all checks, or for indvidual lint checks, the command mentioned in each section respectivley.
+
+To pass linting, no errors or warnings are allowed - ie all rules must be satisfied. It is possible to override and ignore rules, but this should only be done as a last resort, with an accompanying explination.
```
typo on 'explination'

matthew-chirgwin replied:
Thanks Jordan! Good spot, will fix now 👍

JQrdan reviewed (changes requested):
Tooling looks good! Happy to see a CSS linter on there too

matthew-chirgwin committed:
fix: spelling errors/lint vs formatting clarity

- fixes spelling errors in document
- mentions difference between linting and formatting (and what tools are
responsible for what roles)

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

JQrdan commented:
Is there a way to hook in the style lint with prettier in the same way that eslint can be handled by it? I'm sure there's a stylelint plugin that can auto-fix spacing but if we have any more specific style requirements it could be handy for that to all be handled by prettier.

matthew-chirgwin commented:
> Is there a way to hook in the style lint with prettier in the same way that eslint can be handled by it? I'm sure there's a stylelint plugin that can auto-fix spacing but if we have any more specific style requirements it could be handy for that to all be handled by prettier.

I absolutley agree. My thought was to divide and conquer linting responsibilities between a tool for syntax (ie eslint, stylelint) and one for formatting (prettier).

matthew-chirgwin committed:
refactor: tooling used for header checks

- changes tooling used for header management, as proposed approach did
not work nicely

Signed-off-by: Matthew Chirgwin <chirmatt@uk.ibm.com>

nictownsend reviewed (approved):
LGTM

dlabaj reviewed (approved):
Would like test this out with typescript this week.  Was told recently that eslint can be used with typescript looking into that.  Also agree that having prettier as well would be great. 

LGTM

ppatierno commented on docs/Linting.md:
```diff
@@ -0,0 +1,103 @@
+
+| Rule/Ruleset                                                                              |                                       Rationale |                                                         Implemented by | Auto fixes? |
+| ----------------------------------------------------------------------------------------- | ----------------------------------------------: | ---------------------------------------------------------------------: | ----------: |
+| Must have the Strimzi copyright header (in the appropriate format) at the top of the file | Requirement for all code contributed to Strimzi | [license-check-and-add](https://github.com/awjh/license-check-and-add) |         Yes |
```
where is the copyright header format defined? are you going to use the same we have in the Java source code?

matthew-chirgwin replied:
@samuel-hawker kindly provided me with this:
```
Copyright Strimzi authors.
License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
```
I would have this defined in a file, which the tooling then inserts as required (with the right comments)

ppatierno reviewed (approved):
Happy with this. I just asked a question.

scholzj commented:
@JQrdan Are you fine with this? Looks like you requested some changes on this PR.
